Fixed stream close() to end()

geier1993 opened the pull request:
Hey,

I could not compile your extension using 

```
vsce package
```
because
```
yarn run generate
```
already failed with
```
$ ts-node commands/generate.ts && ts-node package.ts                                                                                                                        

/home/geier/misc/dance/node_modules/ts-node/src/index.ts:245
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
package.ts:3:10 - error TS2724: Module '"./commands"' has no exported member 'commands'. Did you mean 'ICommand'?

3 import { commands } from './commands'
```

It turned out that `commands/index.ts` and `README.md` has not been generated completely.
Problem was that only the first `write(...)` was written because the `close()` at the end of the script happened to fast.

The solution was to use `end()` instead of `close()`: [nodejs/node/issues/5631](https://github.com/nodejs/node/issues/5631)

After this change I managed to create and use the extension.

---
I'm using kakoune heavily on Linux but have to switch to Windows for work... Just using vscodevim made me crazy, thank you for providing this extension :)

71 commented:
Ah, glad that you could fix it! I noticed this error on Linux but didn't manage to fix it.

Do you mind if I ask you how your experience is with it for now? I try to use it daily, but have had some problems (most importantly delays when going from normal to insert mode and vice-versa). I also had some compatibility issues with another extension (Trailing Whitespaces), but AFAIK I can't do anything about them.

geier1993 commented:
> Ah, glad that you could fix it! I noticed this error on Linux but didn't manage to fix it.
> 
> Do you mind if I ask you how your experience is with it for now? I try to use it daily, but have had some problems (most importantly delays when going from normal to insert mode and vice-versa). I also had some compatibility issues with another extension (Trailing Whitespaces), but AFAIK I can't do anything about them.

Yet I have not played to much with it and did not use any other extension in combination yet.
 * For some reasons the goto command is not working: `command 'dance.goto' not found`.
 * Search is not working. There is a dialog where I can enter some regexp, but pressing Enter does not execute anything
 * Moving to the end of the last word in a buffer (i.e. pressing e when the cursor is before the last word of an buffer) does not work
 * I had/have to reassign a lot of shortcuts because I am using a dvorak keymap... but this is a general VSCode problem
 * I can not say anything about performance and delays - typing in kakoune is faster than in vscode in general...

So in general I'm having more problems with VSCode, sometimes even the extension host still breaks. That makes me kinda sad...

What I really like is the execution of Javascript with $,$$ and i variables.
Javascript is pretty easy and using this for multiple selection is very helpful :)

71 commented:
- Oops, my bad, I forgot to import the goto commands in `index.ts`. In any case, I haven't implemented all of them yet.
- Search is supposed to search within the current selections, unless we're talking about different things? In any case I could show a quick error message if no results are found.
- I'll look into it.
- I understand, yes. I mainly use AZERTY as well as a custom keyboard layout, and I had the same problem. If you don't mind sharing your config, I might upload some default keybindings to the git that we can just copy/paste to get started on well-known keyboard layouts.
- I see. I haven't really had problems with the extension host on my end though, huh.

I'm glad you liked the idea! It's also the replace "mode" I saw myself using the most.

geier1993 commented:
>     * Oops, my bad, I forgot to import the goto commands in `index.ts`. In any case, I haven't implemented all of them yet.
> 
>     * Search is supposed to search within the current selections, unless we're talking about different things? In any case I could show a quick error message if no results are found.
> 
>     * I'll look into it.
> 
>     * I understand, yes. I mainly use AZERTY as well as a custom keyboard layout, and I had the same problem. If you don't mind sharing your config, I might upload some default keybindings to the git that we can just copy/paste to get started on well-known keyboard layouts.
> 
>     * I see. I haven't really had problems with the extension host on my end though, huh.
> 
> 
> I'm glad you liked the idea! It's also the replace "mode" I saw myself using the most.

 * I'll upload my keysettings once I have set all keys... (where should I place it?)

 * Select/Replace works without problems. Dance.search.selection and search.selection.smart again shows an error `command 'dance.search.selection.smart' not found`.
However dance.search and all the other derivations (extend and backwards) are not working, even when I have text selected.
The window is not closing and not showing anything. Maybe I can debug it one my own once I am more familiar with vscode extensions

 * dance.marks.combineSelections.*  -- commands are not found as well

Up to know I get along quite good with dance.

I have to say that I am new to VSCode. Handling all this windows etc is still unusual for me and I am missing something like ranger.
Do you have any recommendations?

Emacs with MultipleCursors and Kakoune.el is also very pleasant

geier1993 commented:
* save and restoreSelection istn't implemented either?
 * If I want to select a register (dance.register.select) I get the following error:
   `Unable to listen to keyboard events; is an extension overriding the "type" command (e.g VSCodeVim)?`
   Although I have disabled VSCodeVim
 * *IMPORTANT ISSUE*: currently on my system the window is not scrolling with the cursor. I can move out of the view and have to scroll the the cursor manually.

Maybe I start opening Issues. I want these things to work :)

Here are my current shortcuts.
Some are still missing:
 * cursorPageUp/Down on ctrl+f and ctrl+b (these are overloaded with VSCode, hence everyone should add these on his/her own)
 * I removed ctrl+v because I'm not using it and it is overloading the default paste
 * In kakoune I used ctrl+c to add cursors below. Here I added shift+c and alt+shift+c to add a cursor below and above

```
// Place your key bindings in this file to override the defaults
[
    {
        "key": "shift+'",
        "command": "dance.registers.select",
        "when": "dance.mode == 'normal'"
    },
    {
        "key": "/",
        "command": "dance.search",
        "when": "dance.mode == 'normal'" 
    },
    {
        "key": "shift+/",
        "command": "dance.search.extend",
        "when": "dance.mode == 'normal'" 
    },
    {
        "key": "alt+/",
        "command": "dance.search.backwards",
        "when": "dance.mode == 'normal'" 
    },
    {
        "key": "shift+alt+/",
        "command": "dance.search.backwards.extend",
        "when": "dance.mode == 'normal'" 
    },
    {
        "key": "space",
        "command": "dance.selections.clear",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "space",
        "command": "-dance.selections.clear",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "shift+j",
        "command": "dance.down.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "j",
        "command": "-dance.down.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "shift+h",
        "command": "dance.left.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "h",
        "command": "-dance.left.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "shift+l",
        "command": "dance.right.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "l",
        "command": "-dance.right.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "shift+k",
        "command": "dance.up.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "k",
        "command": "-dance.up.extend",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "shift+0",
        "command": "dance.rotate"
    },
    {
        "key": "shift+9",
        "command": "dance.rotate.backwards"
    },
    {
        "key": "shift+7",
        "command": "dance.selections.align"
    },
    {
        "key": "shift+alt+7",
        "command": "dance.selections.align.copy"
    },
    {
        "key": "shift+`",
        "command": "dance.toUpperCase"
    },
    {
        "key": "shift+c",
        "command": "editor.action.insertCursorBelow"
    },
    {
        "key": "alt+shift+c",
        "command": "editor.action.insertCursorAbove"
    },
    {
        "key": "shift+.",
        "command": "dance.indent"
    },
    {
        "key": "shift+,",
        "command": "dance.deindent"
    },
    {
        "key": "ctrl+h",
        "command": "workbench.action.navigateLeft"
    },
    {
        "key": "ctrl+l",
        "command": "workbench.action.navigateRight"
    },
    {
        "key": "ctrl+k",
        "command": "workbench.action.navigateUp"
    },
    {
        "key": "ctrl+j",
        "command": "workbench.action.navigateDown"
    },   
    {
        "key": "ctrl+v",
        "command": "-dance.tmp.insert",
        "when": "editorTextFocus && dance.mode == 'normal'"
    },
    {
        "key": "ctrl+v",
        "command": "-dance.tmp.normal",
        "when": "editorTextFocus && dance.mode == 'insert'"
    }
]
```

71 commented:
To be honest I have mostly used Windows in the past few months, so I haven't played with Kakoune that much for real projects, and I therefore don't know the exact way it behaves in some situations, which is why some details are completely off. Apologies for that.

I'll work on my AZERTY keybindings and upload them. If you finish yours early, I'm thinking about putting them all in a `/layouts` or `/keybindings` directory (if you have other naming suggestions, I'll take them).

As for the search, that's possible. I haven't used it, so I might have not implemented it and forgot about it.

I'm also getting this `type` command error, but I don't know where it originates. I'll try making sure I'm not shooting myself in the foot though.

By the way, thanks for the contribution! There is still quite a bit of work needed to get even close to the Kakoune editing experience, so if you want to add features or modify some behavior, please don't hesitate to open new PRs!

As a side note, what's your opinion on the cursor? Do you prefer a block or a line? I think using a line is better and "more precise," but it makes it less obvious whether we're in normal or insert mode.

geier1993 commented:
I like what you have done so far, and I like the name of the extension ... :P

Step by step we should make it to implement all the missing peaces. 


> As a side note, what's your opinion on the cursor? Do you prefer a block or a line? I think using a line is better and "more precise," but it makes it less obvious whether we're in normal or insert mode.

Right now I'm getting on with the line. The difference of normal/insert mode can be seen at the line number...
What I am curious about is the behaviour of replace character. With a block it is clear which character is going to be replaced, how is it with the line?
Again the same error `Unable to listen to keyboard events; is an extension overriding the "type" command (e.g VSCodeVim)?`

71 commented:
Ahah, thanks! I started with "kakoune-vscode" or something of the sort, but since it's not a full emulation but rather a collection of commands and key bindings, I decided to drop "Kakoune" from the name, and somehow chose the name "Dance."

Yup, there's quite a bit of work and I don't work on it as much as I used to, but I'd like to use it as my daily driver without feeling like it's a compromise at some point.

For me, the "replace character under the cursor" case is the only case (at least on top of my mind) where the block cursor can be useful. In other cases, I think the line delimits better what characters are or aren't in the current selection. In Dance, though, typing `r` without a selection doesn't do anything (the default selection is empty, rather than being the character under the cursor).

> Again the same error `Unable to listen to keyboard events; is an extension overriding the "type" command (e.g VSCodeVim)?`

I have a fix on the way!
